### DISCUSSION TITLE
When TLSv1.3 uses SSL_set_psk_client_callback, the client will reject HRR that select a ciphersuite with SHA384

### DISCUSSION THREAD

**Discussion by ftx456789** [openssl/openssl#25182] [Q&A] [answered]
According to the manual, `SSL_set_psk_client_callback` can only be used with SHA256. In the HRR scenario, `SSL_AD_INTERNAL_ERROR` is generated when the server uses SHA384. 

In my opinion:
When hash is not SHA256, `SSL_set_psk_client_callback' should not be used?

https://github.com/openssl/openssl/blob/21bcae6561d73e629f11e19975f24283559d36c0/ssl/statem/extensions_clnt.c#L1144-L1151

**Comment by mattcaswell** [comment, marked as answer]
Right. `SSL_set_psk_client_callback` was originally designed before TLSv1.3 existed, and is primarily intended for use with TLSv1.2 PSKs. It will *work* in TLSv1.3 but the SHA256 hash is always used. If you need to have more control and use difference hashes then  `SSL_set_psk_use_session_callback` (or `SSL_CTX_set_psk_use_session_callback`) should be used instead as discussed on the man page:

https://docs.openssl.org/master/man3/SSL_CTX_set_psk_client_callback/
